### PR TITLE
Quantized Embedding

### PR DESCRIPTION
**Pull request by mmoffatt2**
Moved quantize embedding code from "Quantization and Binarization Implementations for Linear Layers" to new PR

### CODE REVIEW

File: quantization/quantize.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        # gradient is approximated by simply passing the gradient from the output directly to the input, 
+        # ignoring the quantization operation
+        return grad_output, None, None
+
+quantize_dictionary = {
+    "affine_quant": affine_quantize,
+    "stochastic_quant": stochastic_quantize
+}
+
+_fake_quantize = FakeLinearQuantizationFunction.apply
```
Comment (gkielian) [inline quantization/quantize.py:119]: Maybe remove from this CL

File: model.py [hunk 2]
```diff
@@ -406,7 +416,11 @@ def update_block_size(self, new_block_size):
         # Function to increase block size dynamically
         if new_block_size > self.config.block_size:
             self.config.block_size = new_block_size
-            self.transformer.wpe = nn.Embedding(new_block_size, self.config.n_embd)
+            if self.config.quantize_wpe:
```
Comment (gkielian) [inline model.py:419]: maybe add `and if self.config.use_abs_pos_embeddings`

Comment (gkielian) [reply model.py:419]: There might be some others as well

File: train.py [hunk 3]
```diff
@@ -126,6 +126,11 @@ def parse_args():
     model_group.add_argument( "--linear_mean_init", type=float, default=0.0)
     model_group.add_argument( "--linear_std_init", type=float, default=0.02)
 
+    # Quantization Options
+    model_group.add_argument("--quantize_wte", default=None, action=argparse.BooleanOptionalAction, help="Whether the word embedding is quantized")
+    model_group.add_argument("--quantize_wpe", default=None, action=argparse.BooleanOptionalAction, help="Whether the word position embedding is quantized")
```
Comment (gkielian) [inline train.py:131]: If you can add wte quantization precision and also one for wpe

**Review by gkielian — CHANGES_REQUESTED** [review]
Looks good! Added feedback within single comments

**Comment by mmoffatt2** [comment]
Addressed PR feedback. With wte and wpe quantized, model still converges and has good inference sample output.

image
image

**Review by gkielian — APPROVED** [review]

**Comment by gkielian** [comment]
Looks good! : D 

model.py One step closer to having full quantization.

Merging.

**Pull request merged**